validate player secrets on public API requests

matthewejones opened the pull request:
done:
- Validate player secrets for the following public api requests:
  - `players`
  - `model`
  - `download`
  - `download/text`

- Download buttons use get request

todo:
- send requests with secret in custom header

Resolves #70

ChristophNiehoff commented on src/client/components/sidebar/sidebar.js:
```diff
@@ -20,13 +20,18 @@ class Sidebar extends React.Component {
     this.apiBase = (process.env.NODE_ENV === 'production') ? '/api' : `${window.location.protocol}//${window.location.hostname}:${API_PORT}`;
   }
 
+  url(string) {
```
I think, `string` is a confusing name for a variable. Maybe use `apiEndpoint`?

ChristophNiehoff replied:
And maybe even call the function `apiEndpointUrl`

ChristophNiehoff commented on src/server/publicApi.js:
```diff
@@ -56,19 +56,31 @@ const runPublicApi = (gameServer) => {
+
+    // validate secret
+    if (!isSecretValid(ctx, metadata)) {
+        return;
```
What HTTP code will be returned?  I hope this does not default to 500 or so?

ChristophNiehoff replied:
Or even 200...

ChristophNiehoff commented on src/server/publicApi.js:
```diff
@@ -143,7 +160,13 @@ const runPublicApi = (gameServer) => {
-
+function isSecretValid(ctx, metadata) {
+    if (metadata.players[ctx.params.id].credentials !== ctx.params.secret) {
+        ctx.status = 403;
```
Oh, I see...  the http code is set here...

Actually, I wouldn't expect an `isSomething()` function to have such an side effect. Is there some better place to set the return code?
Maybe wrap the above `if`-blocks in something like an `assertValidSecrets()` function?

matthewejones replied:
I think it's probably clearest if I just set the return code in the `if(!isSecretValid())` statement

ChristophNiehoff commented on src/server/publicApi.js:
```diff
@@ -56,19 +56,31 @@ const runPublicApi = (gameServer) => {
     });
 
-    router.get('/model/:id', async ctx => {
+    router.get('/model/:game/:id/:secret', async ctx => {
```
In general, it is not a good practice to put secrets as query parameters into REST endpoints. The reason is that the full requested URLs are usually written down in the webservers logs and can leak from there...

Could these three numbers be sent in the HTTP body (i.e. in JSON)? Or alternatively, put them into a custom HTTP header as described here: https://www.fullcontact.com/blog/2016/04/29/never-put-secrets-urls-query-parameters/

ChristophNiehoff commented on src/server/publicApi.js:
```diff
@@ -10,8 +10,16 @@ import { getTypeString, escapeMarkdownText, isGameModeCornucopia } from '../util
   const router = new Router();
-  router.get('/players/:id', async (ctx) => {
-    const matchID = ctx.params.id;
+  router.get('/players/:game/:id', async ctx => {
```
I would even assume that whatever library is used to make the GET request in the frontend, it will probably natively support to make requests with basicAuth

ChristophNiehoff replied:
E.g. for superagent you should be able to do something like

request 
  .get('http://local')
  .auth('tobi', 'learnboost')

see https://visionmedia.github.io/superagent/

I would try whether this already sets the Authorization header as wanted.

matthewejones commented on src/client/components/downloadbutton/downloadbutton.js:
```diff
@@ -0,0 +1,75 @@
+      let filename = this.getFilename(res);
+      res.blob().then(fileBlob => {
+        var url = URL.createObjectURL(fileBlob);
+        var a = document.createElement('a');
```
I think we may have talked about this, I couldn't actually get it to download using window.assign() or window.open() it would just display the file on the screen

matthewejones commented:
- I've changed everything to use basic auth now.
- I also added some error handling so instead of the game crashing it will just print errors in the console. For the download buttons I've made them turn red like the copybuttons to give the user some feedback when they don't work
- I changed the copybuttons to use async, await and try{}catch{} instead of promises. I think this makes it more consistent with the rest of the application. I don't know much about promises/async functions so let me know if this is wrong/worse somehow

ChristophNiehoff commented on src/client/components/downloadbutton/downloadbutton.js:
```diff
@@ -0,0 +1,95 @@
+      >
+        <FontAwesomeIcon icon={this.state.icon} fixedWidth /> &nbsp;
+        {this.props.children }
+        {this.state.text}
```
Is `this.state.text` even defined?

matthewejones replied:
no sorry, was thinking about having the button show the error message if it fails but when I tested it, it looked bad because the icon and stuff would all move when you clicked it. Just forgot to get rid of this

ChristophNiehoff commented on src/server/__tests__/server.test.js:
```diff
@@ -75,13 +76,14 @@ it('retrieve the model for a game', async () => {
   expect(response.body).toStrictEqual(model)
 });
 
 it('download the final model for a game', async () => {
```
Could you please also add tests that the download etc. does NOT work for wrong credentials?

ChristophNiehoff replied:
and we should probably test all possibilities:
* wrong password
* wrong username
* not Authorization header
* malformed Authorization header

ChristophNiehoff commented on package.json:
```diff
@@ -7,6 +7,7 @@
     "@fortawesome/fontawesome-svg-core": "^1.2.36",
     "@fortawesome/free-solid-svg-icons": "^5.15.3",
     "@fortawesome/react-fontawesome": "^0.1.15",
+    "basic-auth": "^2.0.1",
```
Great!  It's good to use a library for this and implement it yourself!

ChristophNiehoff commented on src/server/publicApi.js:
```diff
@@ -60,14 +56,37 @@ const runPublicApi = (gameServer) => {
-    const matchID = ctx.params.id;
+  //authorise
+  router.use('/:matchID/', async (ctx, next) => {
+    const user = auth(ctx);
```
Maybe call this `credentials`? My first thought was: "why is the username compared to the password?"

ChristophNiehoff commented on src/server/publicApi.js:
```diff
@@ -60,14 +56,37 @@ const runPublicApi = (gameServer) => {
+    const game = await gameServer.db.fetch(ctx.params.matchID, { metadata: true });
+    const metadata = game.metadata;
+    if(user && metadata && metadata.players && user.pass === metadata.players[user.name].credentials) {
+      return await next();
```
I think it is a great solution writing this middleware!

matthewejones replied:
Thanks!

ChristophNiehoff commented on src/server/publicApi.js:
```diff
@@ -60,14 +56,37 @@ const runPublicApi = (gameServer) => {
-  router.get('/model/:id', async (ctx) => {
-    const matchID = ctx.params.id;
+  //authorise
+  router.use('/:matchID/', async (ctx, next) => {
```
Do I understand it correctly that routes above this `router.use()` call are not authenticated and only the ones belows actually are?

If yes, I am not sure if this is a good solution. This will definitly lead to confusion if someone wants to add another endpoint in the future...  Would it be possible to attach the middleware only to certain routes (and make this explicit)?  Or alternatively, could you add something like a whitelist, such that the middleware by default wants to authorize everything and only the routes in the whitelist are not considered?

matthewejones replied:
It should be that the routes which match the path `'/:matchID/'` use the authorisation.
the syntax is `router.use(path, callback)` where the path is either a string or an array of paths.
It might be better to use `'/:matchID/*'` or to explicitly list all the complete paths: `['/:matchID/players', '/:matchID/model/', ... ]`

ChristophNiehoff replied:
Oh, I see. Cool!  I think `'/:matchID/*'` would be good. Then the logic is that everything specific to a game should be authenticated, which sounds reasonable to me.

Just for clearity, could you move the `router.use()` statement above all the route definitions? And could you please mention in the comment that the middleware only applies to the `'/:matchID/*'` routes, please?

ChristophNiehoff commented:
@matthewejones Just had a manual test of this PR. Works great!  I love it!
And regarding your question: Personally, I prefer async/await over promises as I think that makes the code more readable. So, I am very happy with your change!
